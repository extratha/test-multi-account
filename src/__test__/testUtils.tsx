import { RenderOptions, RenderResult, render } from '@testing-library/react';
import userEvent from '@testing-library/user-event';
import { SWRConfig } from 'swr';
import { theme } from '@/config/config-mui/theme';
import { ThemeProvider as MUIThemeProvider } from '@mui/material/styles';
import { NextIntlClientProvider } from 'next-intl';
import commonTh from '../../public/locales/th/common.json';
import commonEn from '../../public/locales/th/common.json';
type RenderWithProvider = (
  elm: React.ReactElement,
  renderOptions?: ProviderOptions,
) => RenderResult;

export type ProviderOptions = {
  formOptions?: Record<string, unknown>;
} & RenderOptions;

const messages = {
  ...commonTh,
  ...commonEn,
};

const renderWithProviders: RenderWithProvider = (
  component,
  renderOptions?,
  locale = 'th',
) => {
  const wrapper = ({ children }: { children: React.ReactNode }) => (
    <NextIntlClientProvider locale={locale} messages={messages}>
      <MUIThemeProvider theme={theme}>
        <SWRConfig value={{ provider: () => new Map() }}>{children}</SWRConfig>
      </MUIThemeProvider>
    </NextIntlClientProvider>
  );

  return render(component, {
    wrapper,
    ...renderOptions,
  });
};

export { renderWithProviders as render, userEvent };
export * from '@testing-library/react';
