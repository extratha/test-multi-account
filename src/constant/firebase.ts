export const remoteConfigKey = {
  SYMPTOM_CHECKER: "SymptomChecker",
  LAB_INTERPRET_REQUIRE_FIELDS: "LabInterpretRequireFields",
};
