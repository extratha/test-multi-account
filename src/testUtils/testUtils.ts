import { act, render as testRender } from "@testing-library/react";
import dayjs from "dayjs";
import flushPromises from "flush-promises";
import { ReactElement } from "react";

import { TestProvider } from "./TestProvider";

export const API = {
  LOGIN: "/auth/login",
  EXAMPLE_AI_INTERPRET_URL: "/lab/examples/labId",
  GET_LAB_INTERPRETS_ID_URL: "/lab/interprets/labId",
  SUBMIT_LAB_INTERPRETS_URL: "/lab/interprets",
  CONSENT: "/consents",
  TERMS_AND_CONDITIONS: "/consents/terms-conditions",
  TERMS_AND_CONDITIONS_LATEST: "/consents/terms-conditions/latest",
  PRIVACY_POLICY: "/consents/privacy-policies",
  PRIVACY_POLICY_LATEST: "/consents/privacy-policies/latest",
  PUBLIC_TERMS_CONDITIONS: "/consents/terms-conditions/public",
  PUBLIC_PRIVACY_POLICIES: "/consents/privacy-policies/public",
};

export const render = (ui: ReactElement) => {
  return testRender(ui, { wrapper: TestProvider });
};

export const flushPromise = async () => {
  await act(async () => {
    await flushPromises();
  });
};

export const advanceTimersByTime = async (time: number) => {
  await act(async () => {
    jest.advanceTimersByTime(time);
  });
};

export const setMockDate = (date: string) => {
  jest.setSystemTime(dayjs(date).valueOf());
};
