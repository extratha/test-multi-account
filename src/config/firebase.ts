// Import the functions you need from the SDKs you need
import { envConfig } from "@/constant/env";
import { initializeApp } from "firebase/app";

// TODO: Add SDKs for Firebase products that you want to use
// https://firebase.google.com/docs/web/setup#available-libraries

// Your web app's Firebase configuration
// For Firebase JS SDK v7.20.0 and later, measurementId is optional
const firebaseConfig = {
  apiKey: envConfig.firebase.apiKey,
  authDomain: envConfig.firebase.authDomain,
  projectId: envConfig.firebase.projectId,
  storageBucket: envConfig.firebase.storageBucket,
  messagingSenderId: envConfig.firebase.messagingSenderId,
  appId: envConfig.firebase.appId,
  measurementId: envConfig.firebase.measurementId,
};

const firebaseApp = initializeApp(firebaseConfig);
export default firebaseApp;
