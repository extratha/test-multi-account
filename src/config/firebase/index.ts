// Import the functions you need from the SDKs you need
import { initializeApp, getApps } from "firebase/app";
import { getAnalytics } from "firebase/analytics";
import { staticEnvConfig } from "@/constant/env";

// TODO: Add SDKs for Firebase products that you want to use
// https://firebase.google.com/docs/web/setup#available-libraries

// Your web app's Firebase configuration
// For Firebase JS SDK v7.20.0 and later, measurementId is optional
const firebaseConfig = {
  apiKey: staticEnvConfig.firebase.apiKey,
  authDomain: staticEnvConfig.firebase.authDomain,
  projectId: staticEnvConfig.firebase.projectId,
  storageBucket: staticEnvConfig.firebase.storageBucket,
  messagingSenderId: staticEnvConfig.firebase.messagingSenderId,
  appId: staticEnvConfig.firebase.appId,
  measurementId:staticEnvConfig.firebase.measurementId
};

// Initialize Firebase
export const app = getApps().length === 0 ? initializeApp(firebaseConfig) : getApps()[0];

export const analytics = (typeof window !== "undefined") ? getAnalytics(app) : undefined;