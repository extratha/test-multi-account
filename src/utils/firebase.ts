import { fetchAndActivate, getRemoteConfig, getValue } from "firebase/remote-config";

import firebaseApp from "@/config/firebase";
import { DashboardMenuConfigResult, InputGroupConfigResult, SymptomCheckerConfigResult } from "@/types/model.ui";
import { REMOTE_CONFIG_INTERVAL } from "../config";

const getConfig = async (key: string) => {
  const remoteConfig = getRemoteConfig(firebaseApp);
  remoteConfig.settings.minimumFetchIntervalMillis = REMOTE_CONFIG_INTERVAL;
  await fetchAndActivate(remoteConfig);
  return getValue(remoteConfig, key);
};

export const getDashboardMenuConfig = async () => {
  const config = await getConfig("DashboardMenu");
  return JSON.parse(config.asString()) as DashboardMenuConfigResult;
};

export const getLabInterpretFieldsConfig = async () => {
  const config = await getConfig("LabInterpretRequireFields");
  return JSON.parse(config.asString()) as InputGroupConfigResult[];
};

export const getSymptomCheckerConfig = async () => {
  const config = await getConfig("SymptomChecker");
  return JSON.parse(config.asString()) as SymptomCheckerConfigResult;
};
