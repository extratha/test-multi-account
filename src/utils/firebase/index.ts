import { app } from "@/config/firebase";
import { RemoteConfig, fetchAndActivate, getRemoteConfig, getValue } from "firebase/remote-config";

let remoteConfigVar: RemoteConfig;
if (typeof window !== 'undefined') {
	remoteConfigVar = getRemoteConfig(app);
	remoteConfigVar.settings.minimumFetchIntervalMillis = 60000; // 1 Minute
}

export const remoteConfig = {
	getString: (key: string) => fetchAndActivate(remoteConfigVar).then(() => getValue(remoteConfigVar, key).asString()),
	getBoolean: (key: string) => fetchAndActivate(remoteConfigVar).then(() => getValue(remoteConfigVar, key).asBoolean()),
	getNumber: (key: string) => fetchAndActivate(remoteConfigVar).then(() => getValue(remoteConfigVar, key).asNumber())
}