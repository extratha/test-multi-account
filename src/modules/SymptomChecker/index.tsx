"use client";

import { Box, Divider, Paper as MuiPaper, Stack, styled, Typography } from "@mui/material";
import Image from "next/image";
import { useEffect, useRef, useState } from "react";

import { IconAiInterpret, IosDeviceFrame } from "@/assets";
import PageTitle from "@/components/Typography/PageTitle";
import useTranslation from "@/locales/useLocale";
import { SymptomCheckerConfigResult } from "@/types/model.ui";
import { getSymptomCheckerConfig } from "@/utils/firebase";

const Wrapper = styled(Stack)({
  flex: 1,
  width: "100%",
  padding: "24px",
});

const Paper = styled(MuiPaper)({
  flex: 1,
  display: "flex",
  flexDirection: "column",
  borderRadius: "16px",
});

const Header = styled(Stack)({
  padding: "32px 40px 0px",
  "& .MuiDivider-root": {
    marginTop: "8px",
  },
});

const Content = styled(Stack)({
  position: "relative",
  flex: 1,
  padding: "8px 40px",
  overflow: "hidden",
});

const DeviceWrapper = styled(Stack)({
  position: "absolute",
  top: 0,
  left: 0,
  width: "100%",
  height: "100%",
  justifyContent: "center",
  alignItems: "center",
});

const DeviceFrameSection = styled(Box)({
  position: "absolute",
  width: "0px",
  height: "0px",
});

const ContentIframe = styled("iframe")({
  position: "absolute",
  width: "100%",
  height: "100%",
  padding: "16px",
  border: "0px",
});

const DeviceImage = styled(Image)({
  pointerEvents: "none",
});

const initialConfig: SymptomCheckerConfigResult = {
  url: "",
};

const SymptomChecker = () => {
  const { translation } = useTranslation();

  const [isLoading, setIsLoading] = useState(true);
  const [config, setConfig] = useState(initialConfig);

  const refContent = useRef<HTMLDivElement>(null);
  const refDeviceSection = useRef<HTMLDivElement>(null);
  const refContentIframe = useRef<HTMLIFrameElement>(null);

  const calculateDeviceFrame = () => {
    if (!refContent.current || !refDeviceSection.current || !refContentIframe.current) return;
    const contentHeight = refContent.current.offsetHeight;
    const margin = 24;
    const maxHeight = 900;

    let height = contentHeight - margin;

    if (height > maxHeight) height = maxHeight;

    const width = height * 0.5;

    refDeviceSection.current.style.width = `${width}px`;
    refDeviceSection.current.style.height = `${height - margin}px`;
    refContentIframe.current.style.borderRadius = `${height * 0.075}px`;
  };

  const fetchConfigData = async () => {
    setIsLoading(true);
    const remoteConfig = await getSymptomCheckerConfig();
    setConfig(remoteConfig);
    setIsLoading(false);
  };

  useEffect(() => {
    fetchConfigData();
  }, []);

  useEffect(() => {
    if (isLoading) return;

    calculateDeviceFrame();

    window.addEventListener("resize", calculateDeviceFrame);
    return () => {
      window.removeEventListener("resize", calculateDeviceFrame);
    };
  }, [isLoading]);

  return (
    <>
      {!isLoading && (
        <Wrapper>
          <Paper>
            <Header>
              <Stack direction="row" alignItems="center">
                <Box marginRight="12px">
                  <IconAiInterpret />
                </Box>
                <PageTitle>{translation("Common.symptomChecker.title")}</PageTitle>
                <Box flex="1" />
                <Typography variant="bodyMedium">{translation("Common.symptomChecker.subtitle")}</Typography>
              </Stack>
              <Divider />
            </Header>
            <Content ref={refContent}>
              <DeviceWrapper>
                <DeviceFrameSection ref={refDeviceSection}>
                  <DeviceImage src={IosDeviceFrame} fill alt="check" />
                  <ContentIframe ref={refContentIframe} src={config.url} />
                </DeviceFrameSection>
              </DeviceWrapper>
            </Content>
          </Paper>
        </Wrapper>
      )}
    </>
  );
};

export default SymptomChecker;
