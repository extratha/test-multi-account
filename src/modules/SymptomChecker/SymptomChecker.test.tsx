import { flushPromise, render } from "@/__tests__/testUtils";
import SymptomChecker from ".";

import * as RemoteConfig from "firebase/remote-config";

describe("SymptomChecker", () => {
  const renderSymptomChecker = async () => {
    const view = render(<SymptomChecker />);
    await flushPromise();
    return view;
  };

  beforeEach(() => {
    const spyRemoteAsString = jest.fn().mockImplementation(() => {
      return JSON.stringify({ url: "http://web-url" });
    });

    jest.spyOn(RemoteConfig, "getValue").mockReturnValue({ asString: spyRemoteAsString } as any);
  });

  it("should render correctly", async () => {
    const { asFragment } = await renderSymptomChecker();
    expect(asFragment()).toMatchSnapshot();
  });
});
