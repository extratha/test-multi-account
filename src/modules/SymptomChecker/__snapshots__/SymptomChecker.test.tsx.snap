// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SymptomChecker should render correctly 1`] = `
<DocumentFragment>
  .emotion-0 {
  display: flex;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
  -webkit-flex: 1;
  -ms-flex: 1;
  flex: 1;
  padding: 24px;
  overflow: hidden;
}

.emotion-1 {
  background-color: #fff;
  color: unset;
  -webkit-transition: box-shadow 300ms cubic-bezier(0.4, 0, 0.2, 1) 0ms;
  transition: box-shadow 300ms cubic-bezier(0.4, 0, 0.2, 1) 0ms;
  border-radius: 16px;
  box-shadow: none;
  -webkit-flex: 1;
  -ms-flex: 1;
  flex: 1;
  display: flex;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
  overflow: auto;
}

.emotion-2 {
  display: flex;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
  -webkit-flex: 1;
  -ms-flex: 1;
  flex: 1;
}

.emotion-3 {
  display: flex;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
  padding: 32px 40px 0px;
}

.emotion-3 .MuiDivider-root {
  margin-top: 8px;
}

.emotion-4 {
  display: flex;
  -webkit-flex-direction: row;
  -ms-flex-direction: row;
  flex-direction: row;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
}

.emotion-5 {
  margin-right: 12px;
}

.emotion-6 {
  margin: 0;
  font-family: fontFamily;
  font-weight: 700;
  font-size: 36px;
  line-height: 1.5;
  background: linear-gradient(94deg, #0059D6 0%, #028DF5 50%, #00CBDD 100%);
  -webkit-background-clip: text;
  background-clip: text;
  -webkit-text-fill-color: transparent;
  text-fill-color: transparent;
}

.emotion-7 {
  -webkit-flex: 1;
  -ms-flex: 1;
  flex: 1;
}

.emotion-8 {
  margin: 0;
  font-size: 16px;
  font-weight: 500;
  line-height: 1.5;
}

.emotion-9 {
  margin: 0;
  -webkit-flex-shrink: 0;
  -ms-flex-negative: 0;
  flex-shrink: 0;
  border-width: 0;
  border-style: solid;
  border-color: rgba(0, 0, 0, 0.12);
  border-bottom-width: thin;
}

.emotion-10 {
  display: flex;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
  position: relative;
  -webkit-flex: 1;
  -ms-flex: 1;
  flex: 1;
  padding: 8px 40px;
  overflow: hidden;
}

.emotion-11 {
  display: flex;
  -webkit-flex-direction: row;
  -ms-flex-direction: row;
  flex-direction: row;
  position: absolute;
  top: 0;
  left: 0;
  width: 100%;
  height: 100%;
  -webkit-box-pack: center;
  -ms-flex-pack: center;
  -webkit-justify-content: center;
  justify-content: center;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
}

.emotion-12 {
  position: relative;
  width: 0px;
  height: 0px;
}

.emotion-13 {
  position: absolute;
  pointer-events: none;
}

.emotion-14 {
  width: 100%;
  height: 100%;
  padding: 16px;
  border: 0px;
}

.emotion-15 {
  display: inline-flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-box-pack: center;
  -ms-flex-pack: center;
  -webkit-justify-content: center;
  justify-content: center;
  position: absolute;
  box-sizing: border-box;
  -webkit-tap-highlight-color: transparent;
  background-color: transparent;
  outline: 0;
  border: 1px solid #eceff1;
  margin: 0;
  border-radius: 4px;
  padding: 5px 15px;
  cursor: pointer;
  -webkit-user-select: none;
  -moz-user-select: none;
  -ms-user-select: none;
  user-select: none;
  vertical-align: middle;
  -moz-appearance: none;
  -webkit-appearance: none;
  -webkit-text-decoration: none;
  text-decoration: none;
  color: #212121;
  font-family: fontFamily;
  font-weight: 700;
  font-size: 12px;
  line-height: 1.5;
  text-transform: uppercase;
  min-width: 64px;
  -webkit-transition: background-color 250ms cubic-bezier(0.4, 0, 0.2, 1) 0ms,box-shadow 250ms cubic-bezier(0.4, 0, 0.2, 1) 0ms,border-color 250ms cubic-bezier(0.4, 0, 0.2, 1) 0ms,color 250ms cubic-bezier(0.4, 0, 0.2, 1) 0ms;
  transition: background-color 250ms cubic-bezier(0.4, 0, 0.2, 1) 0ms,box-shadow 250ms cubic-bezier(0.4, 0, 0.2, 1) 0ms,border-color 250ms cubic-bezier(0.4, 0, 0.2, 1) 0ms,color 250ms cubic-bezier(0.4, 0, 0.2, 1) 0ms;
  top: 24px;
  left: 100%;
  margin-left: 24px;
  white-space: nowrap;
}

.emotion-15::-moz-focus-inner {
  border-style: none;
}

.emotion-15.Mui-disabled {
  pointer-events: none;
  cursor: default;
}

@media print {
  .emotion-15 {
    -webkit-print-color-adjust: exact;
    color-adjust: exact;
  }
}

.emotion-15:hover {
  -webkit-text-decoration: none;
  text-decoration: none;
  background-color: rgba(156, 39, 176, 0.04);
  border: 1px solid #9c27b0;
}

@media (hover: none) {
  .emotion-15:hover {
    background-color: transparent;
  }
}

.emotion-15.Mui-disabled {
  color: rgba(0, 0, 0, 0.26);
  border: 1px solid rgba(0, 0, 0, 0.12);
}

.emotion-15:hover {
  border: 1px solid #eceff1;
  background-color: rgba(96, 125, 139, 0.04);
}

.emotion-16 {
  display: inherit;
  margin-right: 8px;
  margin-left: -4px;
}

.emotion-16>*:nth-of-type(1) {
  font-size: 20px;
}

.emotion-17 {
  overflow: hidden;
  pointer-events: none;
  position: absolute;
  z-index: 0;
  top: 0;
  right: 0;
  bottom: 0;
  left: 0;
  border-radius: inherit;
}

<div
    class="MuiStack-root emotion-0"
  >
    <div
      class="MuiPaper-root MuiPaper-elevation MuiPaper-rounded MuiPaper-elevation1 emotion-1"
    >
      <div
        class="MuiStack-root emotion-2"
      >
        <div
          class="MuiStack-root emotion-3"
        >
          <div
            class="MuiStack-root emotion-4"
          >
            <div
              class="MuiBox-root emotion-5"
            >
              <svg />
            </div>
            <p
              class="MuiTypography-root MuiTypography-body1 emotion-6"
            >
              Symptom Checker
            </p>
            <div
              class="MuiBox-root emotion-7"
            />
            <p
              class="MuiTypography-root MuiTypography-bodyMedium emotion-8"
            >
              ที่ปรึกษาทางการแพทย์อัจริยะ
            </p>
          </div>
          <hr
            class="MuiDivider-root MuiDivider-fullWidth emotion-9"
          />
        </div>
        <div
          class="MuiStack-root emotion-10"
        >
          <div
            class="MuiStack-root emotion-11"
          >
            <div
              class="MuiBox-root emotion-12"
              style="width: -12px; height: -48px;"
            >
              <img
                alt="check"
                class="emotion-13"
                src="/img.jpg"
              />
              <iframe
                class="emotion-14"
                src="http://web-url"
                style="border-radius: -1.7999999999999998px;"
              />
              <button
                class="MuiButtonBase-root MuiButton-root MuiButton-outlined MuiButton-outlinedSecondary MuiButton-sizeMedium MuiButton-outlinedSizeMedium MuiButton-colorSecondary MuiButton-root MuiButton-outlined MuiButton-outlinedSecondary MuiButton-sizeMedium MuiButton-outlinedSizeMedium MuiButton-colorSecondary emotion-15"
                tabindex="0"
                type="button"
              >
                <span
                  class="MuiButton-icon MuiButton-startIcon MuiButton-iconSizeMedium emotion-16"
                >
                  <svg />
                </span>
                รีโหลดหน้า
                <span
                  class="MuiTouchRipple-root emotion-17"
                />
              </button>
            </div>
          </div>
        </div>
      </div>
    </div>
  </div>
</DocumentFragment>
`;
